### ISSUE TITLE
A piano roll lock ghost nodes button.

### ISSUE THREAD

**Issue by ca3games** [LMMS/lmms#6227]
### Enhancement Summary
Basically an "inking" mode in the piano roll.
I'm mainly a drawing artist, so this idea naturally comes from inking step when doing 2D artwork on a drawing application.
The idea would be to lock everything that isn't under the ghost notes like when you lock the alpha button of a 2D layer on photoshop/Krita.
Basically the only valid input when recording with a keyboard or drawing with a mouse will be the areas inside the ghost notes.

#### Justification
This could help stuff like recording much easier, like you lay out with the mouse some chords or a melodic line, then simply record again on top using a MIDI keyboard and automate that way "inking" the mouse notes with your keyboard, and overriding things like dynamics.
Also, the recording when inking could only progress when notes are only being held, making it even easier to "ink" in a slowly way things like dynamics and notes micro off sync or not being of a perfect lenght, which are like huge deals when recording using a MIDI keyboard.

#### Mockup
A button next to the ghost notes icon, or could be a toggled alpha icon/background on top of the ghost notes button.

**Comment by qnebra** [comment]
If it is for making copy of pattern what was already being "ghost noted" why just not copy and paste it?
I guess from what I understood, enabling it would lock entire pianoroll, except areas with ghost notes. Which would force user to made copy of that pattern. But ghost notes aren't strict thing, but more of guideline. Like showing preview of previous layer in drawing app.

Also, what if user presses wrong note when this mode is active?

**Comment by ca3games** [comment]
> If it is for making copy of pattern what was already being "ghost noted" why just not copy and paste it? I guess from what I understood, enabling it would lock entire pianoroll, except areas with ghost notes. Which would force user to made copy of that pattern. But ghost notes aren't strict thing, but more of guideline. Like showing preview of previous layer in drawing app.
> 
> Also, what if user presses wrong note when this mode is active?

It's more similar to how layers in a 2D app locks the alpha pixels from a layer under it.
So, the ghost notes would be this under layer used to lock the alpha pixels from the current piano roll.

Sorry if It doesnt make sense, is a feature from 2D drawing apps.

**Comment by Monospace-V** [comment]
I think what OP wants is the ability to selectively draw on ghost notes (or record onto them), so keeping some notes and removing some notes, for variable amounts of time possibly shorter than the ghost note itself. I'm not sure if it's worth it given the two current ways to go about it (un-ghost and copy, then delete; or draw manually on top of ghost notes), but the use case is a good point.

**Comment by musikBear** [comment]
I dont know of you misunderstood anything, but you definitively made me get an understanding of this, with your comment 😁
So the feature should in fact be _de-ghosting_ clicked ghost-notes, so these now would be _normal-notes_
Iow, what @ca3games you want to 'avoid' is only dragging current-note to the _length_ of specific selected ghost-note.
Is that a correct understanding of your proposal?

**Comment by allejok96** [comment]
As I understand this is similar to the step recorder feature, but the steps size (and somehow key) is determined by the ghost notes.

Makes me think of the Simply Piano app if anyone are familiar with it. You can play a song live, and it highlights the notes in the score as you play, but when you miss a note it will pause until you've pressed all the correct keys.

It may be hard to boil this down to a good logic. What if notes overlap? What if i did not press one of the keys in the ghost pattern? Will it skip or wait for it? Or what if i press a key multiple times quickly? Is the order of key press important? What will determine when we move to the next note/section of notes?

**Comment by Monospace-V** [comment]
So it's essentially a roundabout way of humanization?

**Comment by allejok96** [comment]
@Monospace-V is right, this is probably overkill and could be achieved by slightly randomizing everything.

*But* I could see the usefulness for a way of recording note velocity. It is hard to get velocity right unless you are actually playing.

A step-recorder-like feature where you play back the recorded notes by pressing any midi key and it registers and replaces the velocity.

**Comment by musikBear** [comment]
@ca3games need to react..
His reference to 'inking' make me think he want ghosts-notes to be 'de-ghosted' if they are clicked with the his de-ghost-tool, but that is just note-length that need editing
imo -also overkill 
Step-by-step Re-capture of velocity may sound (no pun) interesting, but how should that be done with a pc-keyboard?
If user got a MIDI-KB, he already got velocity data.
...It actually right now gave me the idea that LMMS' own virtual KB could re-capture velocity, based on mousepress position.

**Comment by ca3games** [comment]
@musikBear In 2D painting apps you can lock the alpha pixels of an upper painting layer and just paint only the pixels that contain a lower non alpha opaque pixels from a lower layer. It would be very similar but for LMMS piano roll using the ghost notes.

@Monospace-V Not really roundabound, but yes, it would be cool to make easier to humanize a recording, again, this is a feature for helping noobs and begginers, It may seem not useful to people with keyboard skills, but not everyone has keyboard playing skills.

**Comment by allejok96** [comment]
@ca3games in painting apps you use your cursor, thus you can choose where to paint. You'll need to explain more in detail how LMMS would know where to "paint". This suggestion seems to be about MIDI input. With MIDI you don't have a cursor, so input must happen in a systemized way. Either the cursor moves at a constant pace, or it moves a step for each kepress.

@musikBear recording velocity through the on-screen piano is pointless, since you could just use the green vertical lines like usual. I'm talking explicitly about MIDI keyboards. When playing on a real piano it's much easier to play with emotion, you just press harder whenever you feel like it. It's difficult to look at the notes on screen and know which ones should be loud and which should be quiet. Now, the problem is I'm a bad piano player, so it takes 50 tries before I can play a short section without errors. If I could somehow play awfully and LMMS would just autocorrect me based on the ghost notes, it would be awesome. I think this is what ca3games is saying.

TL;DR the important thing is velocity. So don't overthink it.

**Comment by ca3games** [comment]
@allejok96 No, my idea would be to easily automate the re recording of MIDI notes in the piano roll, using a MIDI keyboard.
Is meant to be used by people wihout strong piano skills (begginers).
Now that I think about could also be some kind of piano roll "autotune", by simply avoiding mistakes when re recording using a MIDI keyboard.

**Comment by qnebra** [comment]
But then, if you want copy of pattern, why to not just copy & paste it? Your feature, from what I understood, would be nice to learn how to play something on piano, it would turn LMMS into kind of "Learn Your Piano" when enabled.

But, but, point with recording multiple takes was with slight differences between each take, small imperfections. If user was more skilled in musicology even changing melody, you know, different harmonies etc.

**Comment by allejok96** [comment]
![bild](https://user-images.githubusercontent.com/7693838/144300737-dbbf6580-9b66-48c7-bb63-0fa74ae146a6.png)

Imagine re-recording this. It would become very complicated. I mean, how would you do it? Your idea was good @ca3games, but as you can see it gets hard to turn into raw logic.

- **Imperfect start/end** is easily fixed by #1165.
- **Velocity** - I can think of simpler ways to record it, that does not involve ghost notes...
- **Piano practice** - nice, but someone must create the notes to begin with. And it is probably better to have a separate program dedicated for piano practicing instead of integrating it in LMMS.
- **Auto tune** - it would be really nice if we could lock to a *key/scale* instead of a ghost pattern.

**Comment by musikBear** [comment]
A lot of different and not all that related input here now..
for me ghostnotes are only rarely used to 'paint' where the actual ghosts _are placed_, but instead the ghosts are meant to be harmonized or kontrapunkted. 
That is a creative process, and lmms cant be expected to do that

**Comment by allejok96** [comment]
Sorry guys for creating lot of off topic confusion. I think I finally understand it now.

> The only valid input when recording with a keyboard or drawing with a mouse will be the areas inside the ghost notes. This could help stuff like recording much easier ... overriding things like dynamics.

So, only accept clicking/drawing/recording [on top of](https://youtu.be/WQjbmJF6tIo?t=26) notes ghost notes. Why? [Humanization](https://github.com/LMMS/lmms/issues/1165).

@ca3games I think you have been mislead by your visual design mind when designing this idea. The feature you have described is basically an AND operation of notes - only create note where normal note AND ghost note overlap.

I suggest this be made an **one time operation**, like the quantize button. Let's imagine we add a few like Union (OR), Difference, Intersection (AND), Exclusion (XAND), similar to the ones in Inkscape. The user could drag a group of selected notes on top of some unselected notes, and then hit the Intersect button. Some advantages of this approach would be:

1. Wider use-case - does not require recording
2. More obvious purpose - "why would you lock ghost notes?"
3. Ghost notes remains used as intended - "harmonized and kontrapunkted"
4. Easier to program - no need to implement "only record when notes are being held"
